lib/Echidna/ABI.hs

```diff
@@ -241,20 +233,37 @@ shrinkAbiCall = traverse $ traverse shrinkAbiValue
 
 -- | Given an 'AbiValue', generate a random \"similar\" value of the same 'AbiType'.
 mutateAbiValue :: MonadRandom m => AbiValue -> m AbiValue
-mutateAbiValue (AbiUInt n x)         = AbiUInt n         <$> mutateNum x
-mutateAbiValue (AbiInt n x)          = AbiInt n          <$> mutateNum x
-mutateAbiValue (AbiAddress _)        = genAbiValue AbiAddressType
+mutateAbiValue (AbiUInt n x)         = getRandomR (0, 9 :: Int) >>= -- 10% of chance of mutation
+                                          \case  
+                                            0 -> (AbiUInt n <$> mutateNum x)
+                                            _ -> return $ AbiUInt n x
+mutateAbiValue (AbiInt n x)          = getRandomR (0, 9 :: Int) >>= -- 10% of chance of mutation
+                                          \case  
+                                            0 -> (AbiInt n <$> mutateNum x)
+                                            _ -> return $ AbiInt n x 
+
+mutateAbiValue (AbiAddress x)        = return $ AbiAddress x
 mutateAbiValue (AbiBool _)           = genAbiValue AbiBoolType
-mutateAbiValue (AbiBytes n b)        = AbiBytes n        <$> addChars getRandom b
-mutateAbiValue (AbiBytesDynamic b)   = AbiBytesDynamic   <$> mutateBS b
-mutateAbiValue (AbiString b)         = AbiString         <$> mutateBS b
-mutateAbiValue (AbiArray n t l)      = AbiArray n t      <$> traverse mutateAbiValue l
-mutateAbiValue (AbiArrayDynamic t l) = AbiArrayDynamic t <$> mutateV t l
+mutateAbiValue (AbiBytes n b)        = do fs <- replicateM n getRandom
+                                          xs <- mutateBS (Just n) fs b 
+                                          return (AbiBytes n xs)
+
+mutateAbiValue (AbiBytesDynamic b)   = mutateBS Nothing [] b >>= (return . AbiBytesDynamic)
+mutateAbiValue (AbiString b)         = mutateBS Nothing [] b >>= (return . AbiString)
+mutateAbiValue (AbiArray n t l)      = do fs <- replicateM n $ genAbiValue t
+                                          xs <- mutateV (Just n) fs (V.toList l) 
```

Comment:
doesn't `Just n` accomplish the same as `Nothing`?

Reply:
`Just n` means the mutated array/string should have size `n`. `Nothing` means the mutated array/string can have any size.